create `gh project` command

mntlty opened the pull request:
Closes https://github.com/cli/cli/issues/7362

Verified with tests, in the terminal and with a testing script, everything is still working 🤞 

Changes from https://github.com/github/gh-projects:
- import `project` as a root cmd
- rename from `projects` to `project` in examples
- use cmd util, especially the native CLI IO packages, rather than the go-gh ones
- use CLI browser 
- add more cmd tests, for flags and required / exclusive arguments

TODO:
- [x] Inherit HTTPClient from command Factory
- [x] Avoid Survey prompts when stdin or stdout are not a tty
- [x] Avoid printing table headers when table output is not a tty
- [x] Avoid printing blank values as "-" and use an empty string instead
- [x] Avoid using table printer for printing one-line confirmation messages
- [x] Avoid using table printer for printing JSON output
- [x] Avoid varying the number of table columns depending on input flags to a command
- [x] Add support for choosing non-github.com instances with GH_HOST

mntlty committed:
copy gh-projects into pkg/cmd/project

mntlty committed:
include project cmd in root cmd

mntlty committed:
rename gh projects to gh project in examples

mntlty committed:
switch cli/go-gh/v2/pkg/tableprinter for cli/cli/v2/internal/tableprinter

mntlty committed:
use internal browser

mntlty committed:
test project cmd flags and input values

mntlty committed:
attempt to fix tests

mislav commented on pkg/cmd/project/close/close.go:
```diff
@@ -0,0 +1,158 @@
+	opts := closeOpts{}
+	closeCmd := &cobra.Command{
+		Short: "Close a project",
+		Use:   "close [number]",
```
We use `<number>` to denote placeholders in usage synopsis. If the argument is optional, that is indicated with square brackets: `[<number>]`

mntlty replied:
it is optional, because the user will be prompted for it if they don't supply it. will update the code

mntlty replied:
updated in bca04ca6e0dbd148abe3561429b4a2ffe550b600

mislav commented on pkg/cmd/project/close/close.go:
```diff
@@ -0,0 +1,158 @@
+		Short: "Close a project",
+		Use:   "close [number]",
+		Example: `
+# close project 1 owned by user monalisa
```
Protip: you can use `heredoc.Doc` to indent embedded string literals together with your Go code and have the string contents be dedented at runtime

mntlty replied:
this one hasn't bothered me - if you prefer to use heredoc I can definitely make the change, lmk!

mislav commented on pkg/cmd/project/close/close.go:
```diff
@@ -0,0 +1,158 @@
+		},
+	}
+
+	closeCmd.Flags().StringVar(&opts.userOwner, "user", "", "Login of the user owner. Use \"@me\" for the current user.")
+	closeCmd.Flags().StringVar(&opts.orgOwner, "org", "", "Login of the organization owner.")
```
I'm seeing both `--user` and `--org` flags supported across multiple commands, but they are both used to supply an owner. If the concept of "owner" is interchangeable and it doesn't matter whether it's a User or an Organization, could we merge the functionality of both flags under `--owner <handle>` or `--owner @me`? (We would keep old flags for backwards compatibility but hide them from docs)

mntlty replied:
@mislav this is a difficult part of the GraphQL apis, which I haven't found a good solution for. Projects are identified by owner, which can be a user or an org, and a number. For example, https://github.com/users/mntlty/projects/1 or https://github.com/orgs/cli/projects/1

Even though the login names are unique, in GraphQL these are separate objects, https://docs.github.com/en/graphql/reference/objects#user and https://docs.github.com/en/graphql/reference/objects#organization, both of which have a `projectV2` field and is the only way I know to specify the type of the owner.

To further complicate things, the graphql package uses struct tags for these mappings, which isn't something that can be dynamically changed at runtime as far as I know.

As a result, there is a lot of code that is almost exactly the same except for the owner (and related go struct) - https://github.com/cli/cli/blob/e2c531a349270014ac3e7dfa70c1c9e057ebfbd3/pkg/cmd/project/shared/queries/queries.go#L1002 is a good example of this.

Finally, even for a project which is owned by a user, they can invite another user or give Actions access as a collaborator, so it's not possible to assume that the owner is always the current user.

In previous iterations there was a flag for login name, and a flag for the owner type - this ended up feeling cleaner. I wish I had a better idea for how to simplify this and not have the duplication everywhere.

mislav commented on pkg/cmd/project/copy/copy.go:
```diff
@@ -0,0 +1,171 @@
+		},
+	}
+
+	copyCmd.Flags().StringVar(&opts.sourceUserOwner, "source-user", "", "Login of the source user owner. Use \"@me\" for the current user.")
+	copyCmd.Flags().StringVar(&opts.sourceOrgOwner, "source-org", "", "Login of the source organization owner.")
+	copyCmd.Flags().StringVar(&opts.targetUserOwner, "target-user", "", "Login of the target organization owner. Use \"@me\" for the current user.")
+	copyCmd.Flags().StringVar(&opts.targetOrgOwner, "target-org", "", "Login of the target organization owner.")
```
Could we replace all these flags with just `--target <handle>` and `--source <handle>`?

mntlty replied:
Copying can be done across users and orgs. For example, if I'm a member of the CLI org, I could copy a project which my user owns to the CLI, or vice versa. We also need to know the type of both the source and destination (more details in https://github.com/cli/cli/pull/7375#discussion_r1190486710)

The alternative is to have something like `--source` , `--target`, `--source-owner-type`, and `--target-owner-type`, which didn't feel clearer.

mislav commented on pkg/cmd/project/copy/copy.go:
```diff
@@ -0,0 +1,171 @@
+
+	_ = copyCmd.MarkFlagRequired("title")
+	copyCmd.MarkFlagsMutuallyExclusive("source-user", "source-org")
+	copyCmd.MarkFlagsMutuallyExclusive("target-user", "target-org")
```
Protip: we have our own "MutuallyExclusive" helper (which is better than Cobra's) in `cmdutil`

mntlty replied:
fixed in d742d33ef817a4404520ed0d267a87b6bf488794

mislav commented on pkg/cmd/project/create/create.go:
```diff
@@ -0,0 +1,140 @@
+	"fmt"
+
+	"github.com/cli/cli/v2/pkg/cmdutil"
+
```
Nit: are these empty lines between different non-stdlib imports intended?

mntlty replied:
nope! I thought my editor was doing that, turns out I'm wrong. Fixed in 7fe71b11da0dc6076efaa3d33651bc7336113e66

mislav commented on pkg/cmd/project/create/create.go:
```diff
@@ -0,0 +1,140 @@
+# add --format=json to output in JSON format
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client, err := queries.NewClient()
```
Could the API client be inherited from the `cmdutil.Factory` instance?

mntlty replied:
I think I had looked into doing that, and it ended up being a pretty significant change because I'm using the github.com/cli/go-gh/v2/pkg/api `GraphQLClient` and it's `Query` method https://github.com/cli/go-gh/blob/v1.2.1/pkg/api/client.go#L138 

which isn't the same as what's available in cmdutil.Factory

mislav commented on pkg/cmd/project/edit/edit.go:
```diff
@@ -0,0 +1,182 @@
+				opts:   opts,
+			}
+
+			if config.opts.visibility != "" && config.opts.visibility != projectVisibilityPublic && config.opts.visibility != projectVisibilityPrivate {
+				return fmt.Errorf("visibility must match either PUBLIC or PRIVATE")
```
Protip: check out `cmdutil.StringEnumFlag` for defining flags which only allow a set of static values

mntlty replied:
I don't think it's a good fit in this case, as visibility is an option and not a required value. The code change would look like

```go
cmdutil.StringEnumFlag(editCmd, &opts.visibility, "visibility", "", "", []string{"", "PUBLIC", "PRIVATE"}, "Update the visibility of the project.")
```

and the resulting help output looks like

```
--visibility string    Update the visibility of the project.: {|PUBLIC|PRIVATE}
```
I think the optional empty string is very hard to notice here

mntlty replied:
I did find a use for it in 417ee119 though!

mislav commented on pkg/cmd/project/field-list/field_list.go:
```diff
@@ -0,0 +1,168 @@
+	listCmd.Flags().StringVar(&opts.userOwner, "user", "", "Login of the user owner. Use \"@me\" for the current user.")
+	listCmd.Flags().StringVar(&opts.orgOwner, "org", "", "Login of the organization owner.")
+	listCmd.Flags().StringVar(&opts.format, "format", "", "Output format, must be 'json'.")
+	listCmd.Flags().StringVar(&opts.limit, "limit", "", "Maximum number of fields. Defaults to 100. Set to 'all' to list all fields.")
```
We usually default to 30 per page for most commands, especially since most people's terminals aren't that tall by default and a more conservative limit matches both web output and yields faster API responses. Of course, based on the data set you're interested in, you can choose to have a higher default limit if it makes sense, but consider lowering it for most commands.

mntlty replied:
I selected 100 based on https://docs.github.com/en/graphql/overview/resource-limitations#node-limit 
fixed in 0c9651aa

mntlty replied:
6c89a5bd also updated this code

mislav commented on pkg/cmd/project/field-list/field_list.go:
```diff
@@ -0,0 +1,168 @@
+
+func printResults(config listConfig, fields []queries.ProjectField, login string) error {
+	if len(fields) == 0 {
+		config.tp.AddField(fmt.Sprintf("Project %d for login %s has no fields", config.opts.number, login))
```
It's a misuse of tableprinter to print a single error message to stdout and immediately exit.

For these kinds of cases, your command should return an `cmdutil.NoResultsError` instead.

mntlty replied:
agreed! sadly this wasn't exposed in a way that I could use it when making the extension, and forgot to update when making the move.

fixed in 9e2e8ac4

mislav commented on pkg/cmd/project/item-add/item_add.go:
```diff
@@ -0,0 +1,157 @@
+
+	addItemCmd.Flags().StringVar(&opts.userOwner, "user", "", "Login of the user owner. Use \"@me\" for the current user.")
+	addItemCmd.Flags().StringVar(&opts.orgOwner, "org", "", "Login of the organization owner.")
+	addItemCmd.Flags().StringVar(&opts.itemURL, "url", "", "URL of the issue or pull request to add to the project. Note that the name of the owner is case sensitive, and will fail to find the item if it does not match. Must be of form https://github.com/OWNER/REPO/issues/NUMBER or https://github.com/OWNER/REPO/pull/NUMBER")
```
The size of this flag's description raises alarm bells for me. Could it just be "URL of the issue or pull request to add to the project", with everything else being handled as an implementation detail, and the user not necessarily be aware of it?

mntlty replied:
it could be - a user had opened an issue in the extension with this, because while https://github.com/cli/cli and https://github.com/CLI/cli will resolve in github.com, they won't both resolve in https://docs.github.com/en/graphql/reference/queries#resource

to avoid confusion I added a long disclaimer, if you feel it's more trouble than it's worth it can definitely be removed

mislav commented on pkg/cmd/project/item-add/item_add.go:
```diff
@@ -0,0 +1,157 @@
+
+func printResults(config addItemConfig, item queries.ProjectItem) error {
+	// using table printer here for consistency in case it ends up being needed in the future
+	config.tp.AddField("Added item")
```
Ditto: not convinced that tableprinter is a good mechanism to print anything other than structured data that spans multiple columns.

mntlty replied:
what would you suggest changing it to?

mislav commented on pkg/cmd/project/item-edit/item_edit.go:
```diff
@@ -0,0 +1,286 @@
+# add --format=json to output in JSON format
+
+# edit a draft issue title and body
+gh project item-edit --id DRAFT_ISSUE_CONTENT_ID --title "a new title" --body "a new body"
```
Since every invocation of `item-edit` has to pass the required `--id` value as it uniquely select the item to edit, could it be received as required argument instead? E.g.
```sh
gh project item-edit <item-ID> --field-id <field> --project-id <project> --text "new text"
```
We do this in other gh commands: for instance, it's `gh issue close 123`, not `gh issue close --number 123`.

mntlty replied:
this felt more consistent with other commands that take the project number as a positional argument, for example https://github.com/cli/cli/blob/9e2e8ac441cf4fc3fc3414e929893966d37a7e9d/pkg/cmd/project/item-add/item_add.go#L42

what do you think?

mislav commented on pkg/cmd/project/item-edit/item_edit.go:
```diff
@@ -0,0 +1,286 @@
+	editItemCmd.Flags().StringVar(&opts.itemID, "id", "", "ID of the item to edit (required). For draft issues, the ID is for the draft issue content which is prefixed with `DI_`. For other issues, it is the ID of the project item.")
+	editItemCmd.Flags().StringVar(&opts.format, "format", "", "Output format, must be 'json'.")
+
+	editItemCmd.Flags().StringVar(&opts.title, "title", "", "DRAFT ISSUE - Title of the draft issue item to edit.")
+	editItemCmd.Flags().StringVar(&opts.body, "body", "", "DRAFT ISSUE - Body of the draft issue item to edit.")
```
My mind boggles at this "draft issue" feature. Looking at its entire doc string and the flag set, it almost feels like `item-edit` is a bit overloaded. But short of changing that (I wouldn't suggest it at it point), could we do something to make the usage of this clearer? I'm afraid I'm not familiar with the "draft issue" feature of Projects.

mntlty replied:
Draft Issues are items that exist only a project - they do not have their own URL like an Issue or a PR (which is created then added to a Project). However, they can have values such as a number or an assignee, in the same way that Issues or PRs on a Project can.

I agree that the flags for this command are a bit of a mess.  I had a conversation with @samcoe about it here https://github.com/github/gh-projects/issues/84 while I was working on it. I also tried out flag groups in cobra to try to group the draft flags and the non-draft flags, without success.

If you have any ideas for how to improve this, please let me know.

mislav commented on pkg/cmd/project/item-edit/item_edit.go:
```diff
@@ -0,0 +1,286 @@
+	// update draft issue
+	if config.opts.title != "" || config.opts.body != "" {
+		if !strings.HasPrefix(config.opts.itemID, "DI_") {
+			return errors.New("ID must be the ID of the draft issue content which is prefixed with `DI_`")
```
Protip: for any errors related to flags and values passed by flags (or arguments), use `cmdutil.FlagErrorf` to construct an error and return that. The resulting error message will be rendered together with the usage synopsis for the command.

mntlty replied:
fixed in 4dbc0171

mntlty replied:
and also in 38b15a94

mislav commented on pkg/cmd/project/item-edit/item_edit.go:
```diff
@@ -0,0 +1,286 @@
+		config.tp.AddField(fmt.Sprintf("%d", item.Number()))
+	}
+	if item.Repo() == "" {
+		config.tp.AddField(" - ")
```
Is this a placeholder for values that are missing? Could we consider just leaving it empty?

mntlty replied:
when printing rows having an empty value seemed clearer to me, particularly across columns. it can definitely be changed if you don't agree!

mislav commented on pkg/cmd/project/list/list.go:
```diff
@@ -0,0 +1,233 @@
+	listCmd.Flags().BoolVarP(&opts.closed, "closed", "c", false, "Show closed projects.")
+	listCmd.Flags().BoolVarP(&opts.web, "web", "w", false, "Open projects list in the browser.")
+	listCmd.Flags().StringVar(&opts.format, "format", "", "Output format, must be 'json'.")
+	listCmd.Flags().StringVar(&opts.limit, "limit", "", "Maximum number of projects. Defaults to 100. Set to 'all' to list all projects. Note that closed projects are filtered from the final results without the --closed flag.")
```
This is a very long flag description, plus:
- We typically allow the shorthand `-L` wherever there is `--limit`
- We deliberately do _not_ allow unlimited listing (i.e. `"all"`). Users are required to pass a very large number to opt into getting many pages of results.
- The note about closed projects is not necessary. I would expect only to see open projects in the list unless I passed `--closed`.

mntlty replied:
addressed in 6c89a5bd

mntlty replied:
I remember now why I added `closed` to the help output (after a bunch of repeat debugging lol). say for example I run the following:

```
project list --org cli --format=json --limit=3 
```

if any of the projects are closed, they are filtered from the list. I have supplied the limit flag of 3, but may anywhere between 0 and 3 results, depending on the order returned by the API. To make this more confusing, the json output has a field `totalCount` which is the sum of open and closed projects, so a user may be confused wondering why we don't return all of the results to them.

mislav commented on pkg/cmd/project/view/view.go:
```diff
@@ -0,0 +1,208 @@
+		sb.WriteString(fmt.Sprintf("%s (%s)\n\n", f.Name(), f.Type()))
+	}
+
+	// TODO: respect the glamour env var if set
+	out, err := glamour.Render(sb.String(), "dark")
```
Protip: You can use `pkg/markdown.Render` to render markdown instead of going straight to Glamour. You can read the preferred terminal theme from `IOStreams.TerminalTheme()`

mntlty replied:
fixed in f484b822

mislav reviewed (changes requested):
Thanks for the gargantuan effort! This looks great already. My suggestions come down to largely polishing the UI of the final product.

Regarding your questions:

- I don't know yet what we can do about the current `--format json` support re: bringing that in line with how `--json` flag works in the rest of gh. Reworking the current code to support `--json field1,field2` and only conditionally querying those specific fields from the API might be infeasible for now. Let me think about this for a day or two.

- The error I get when I hadn't requested the projects OAuth scope is indeed pretty gnarly. I will look on my end at how that can be handled a bit better.

mislav committed:
Merge remote-tracking branch 'origin/trunk' into mntlty/gh_project

mislav committed:
Depend on github.com/cli/shurcooL-graphql

mislav committed:
Fix command name alignment in `gh project` help docs

mntlty committed:
replace [number] with [<number>]

mntlty committed:
Merge branch 'mntlty/gh_project' of github.com:cli/cli into mntlty/gh_project

mntlty committed:
use cmdutil.MutuallyExclusive

mntlty committed:
whitespace in imports

mntlty committed:
lower default limit from 30 to 100

mntlty committed:
use  cmdutil.NewNoResultsError

mntlty committed:
make limit explicit

mntlty committed:
use markdown.Render

mntlty committed:
update tests

mntlty committed:
use cmdutil.StringEnumFlag

mntlty committed:
use cmdutil.FlagErrorf

mntlty committed:
another spot

mntlty commented:
@mislav I think I've addressed all the feedback so far, please let me know what you think!

mntlty committed:
remove projects alias

mntlty committed:
Merge branch 'trunk' into mntlty/gh_project

mntlty committed:
copy file from trunk

mislav committed:
Switch to githubv4 types to try appease CodeQL

mislav committed:
go mod tidy

mislav committed:
Merge remote-tracking branch 'origin' into mntlty/gh_project

mislav committed:
Change project number to int32 to match the GraphQL Int type

mislav committed:
Simplify project flags

mislav committed:
Simplify project command examples

This mostly focuses on removing examples that should be self-evident after
scanning the generated flags help text.

mislav committed:
Restructure projects API client to reduce coupling to api.NewGraphQLClient and Survey

mislav committed:
:nail_care:

mislav committed:
WIP remove or improve tableprinter

- Avoid printing confirmation messages through tableprinter
- Avoid printing header rows when stdout is redirected

mntlty committed:
replace --owner and --user with a single --login flag

mntlty committed:
missing env var

mntlty committed:
Avoid using table printer for printing one-line confirmation messages

mntlty committed:
Avoid Survey prompts when stdin or stdout are not a tty

mntlty committed:
Avoid printing confirmation message when STDOUT is not a TTY

mislav committed:
Have projects commands inherit the HTTP client from Factory

mntlty committed:
use config.io.Out for writing JSON

mntlty committed:
rename --login to --owner

MarcPow commented:
@williammartin / @mislav - What's the status on this PR?

mislav commented:
@MarcPow Still applying some UI polish and the plan is to be merged to `trunk` within a week.

mntlty committed:
remove description and url from project list output

mntlty committed:
number before title

mntlty committed:
do not truncate IDs or Numbers

mislav committed:
:fire: tableprinter

mislav reviewed (approved):
Thanks for the hard work, @mntlty!!

charlyBro commented:
G
